src/libsystemd-network/dhcp-option.c

```diff
@@ -356,10 +349,168 @@ static int parse_options(const uint8_t options[], size_t buflen, uint8_t *overlo
+ * architecture.
+ *
+ * This function will merge DHCP options by concatenating options of the same type, producing a new
+ * TLV buffer, but using u16 len fields instead of u8. This works due to a DHCP message size having
+ * an inherent upper bound, lower than u16::MAX.
+ *
+ * Basically, this produces the aggregated option buffer the RFC talks about.
+ */
+static int dhcp_option_merge(const DHCPMessage *message, size_t buflen, uint8_t **ret_merged) {
+        _cleanup_ordered_hashmap_free_ OrderedHashmap *opts;
+        uint8_t *aggregate, overload = 0;
+        size_t tot_len = 0;
+        int r = 0;
+
+        opts = ordered_hashmap_new(&merge_option_node_ops);
+        if (!opts)
+                return -ENOMEM;
+
+        /* RFC3396 5. The Aggregate Option Buffer
+         *
+         *   [...], an option that doesn't fit into one field can't overlap the boundary into
+         *   another field - the encoding agent must instead break the option into two parts and
+         *   store one part in each buffer.
+         *
+         *   To simplify this discussion, we will talk about an aggregate option buffer, which will
+         *   be the aggregate of the three buffers. [...] The aggregate option buffer is made up of
+         *   the optional parameters field, the file field, and the sname field, in that order.
+         *
+         * Since options cannot overlap from one field to another, it is safe to read them in
+         * isolation.
+         */
+        r = dhcp_option_merge_append(message->options, buflen, opts, &tot_len, &overload);
+        if (r < 0)
+                return r;
+        if (overload & DHCP_OVERLOAD_FILE) {
+                r = dhcp_option_merge_append(message->file, sizeof(message->file), opts, &tot_len, NULL);
+                if (r < 0)
+                        return r;
+        }
+        if (overload & DHCP_OVERLOAD_SNAME) {
+                r = dhcp_option_merge_append(message->sname, sizeof(message->sname), opts, &tot_len, NULL);
+                if (r < 0)
+                        return r;
+        }
+
+        /* Consolidate in a new buffer */
+        buflen = tot_len + ordered_hashmap_size(opts) * 3;
+        aggregate = new(uint8_t, buflen);
+        if (!aggregate)
+                return r;
+
+        size_t offset = 0;
+        MergeOptionNode *head;
+        ORDERED_HASHMAP_FOREACH(head, opts) {
+                uint8_t code = head->data[0];
+                uint8_t *opt_base = &aggregate[offset];
+                offset += 3;
+
+                size_t len = 0;
+                LIST_FOREACH(options, opt, head) {
+                        size_t optlen = opt->data[1];
+                        memcpy_safe(&aggregate[offset], &opt->data[2], optlen);
+                        offset += optlen;
+                        len += optlen;
+                }
+                opt_base[0] = code;
+                unaligned_write_be16(&opt_base[1], len);
```

Comment:
shouldn't there be an overflow check first?

Reply:
In theory, yes. In practice, even with option splitting, we'll still be limited with the practical maximum size of a DHCP message. It's limited anyways to MTU, with the upper hard limit of a fragmented UDP packet, which including the rest of the DHCP message, will make our options _always_ fit in 64k.

I added the check as an `assert` as exceeding 64k bytes must be a logic error.

I'm leaving this conversation open should you prefer a `return -EINVAL` instead.

Reply:
Welp the fuzzer prefers `return -EINVAL`, so EINVAL it is.